[clang] Refine handling of C++20 aggregate initialization

offsetof opened the pull request:
* Move parts of `InitializationSequence::InitializeFrom` corresponding to C++ [dcl.init.general] p16.6.1 and p16.6.2 into a separate function,`TryConstructorOrParenListInitialization`
* Use it in `TryListInitialization` to implement [dcl.init.list] p3.2
* Fix parenthesized aggregate initialization being attempted in copy-initialization contexts or when the constructor call is ambiguous  
* Fix #88089

cor3ntin commented:
I don't think this resolves the original issue.
Part of the problem is that we are missing a diagnostic saying the implicit constructor is deleted.

cor3ntin reviewed (commented):
This change needs a release note.
Please add an entry to `clang/docs/ReleaseNotes.rst` in the section the most adapted to the change, and referencing any Github issue this change fixes. Thanks!

offsetof committed:
[clang] Refine handling of C++20 aggregate initialization

* Move parts of `InitializationSequence::InitializeFrom` corresponding
  to C++ [dcl.init.general] p16.6.1 and p16.6.2 into a separate
  function, `TryConstructorOrParenListInitialization`
* Use it in `TryListInitialization` to implement [dcl.init.list] p3.2
* Fix parenthesized aggregate initialization being attempted in
  copy-initialization contexts or when the constructor call is ambiguous

cor3ntin commented:
@offsetof CI got confused. Can you try to push an empty commit? Sorry about that

cor3ntin reviewed (approved):
Thanks!

cor3ntin commented:
@offsetof Will you need me to merge that for  you?

cor3ntin committed:
Merge branch 'main' into p0960
